File: src/internal/domain/usecases/user_usecases/log_in_usecase.rs
use std::sync::Arc;

use crate::{
    dto::{
        req::user_req::LogInReq,
    }, internal::domain::{
        repository_interface::user_repository::UserRepository,
        service::password_hash::verify_password, utils::jwt::generate_token::generate_token
    }
};

pub struct LogInUsecase {
    user_repo: Arc<dyn UserRepository + Send + Sync>,
    jwt_secret_key: String,
}

impl LogInUsecase {
    pub fn new(user_repo: Arc<dyn UserRepository + Send + Sync>, jwt_secret_key: String) -> Self {
        Self { user_repo, jwt_secret_key }
    }

    pub async fn log_in_usecase(&self, log_in_req: LogInReq) -> Result<String, String> {
        let check_user_result = self.user_repo.find_by_email(log_in_req.email.clone()).await?;
        if !check_user_result {
            return Err(format!("User not found"));
        }

        let hashed_password_with_salt = self.user_repo.get_hashed_password_with_salt(log_in_req.email.clone()).await?;

        let verify_password_result = verify_password(&log_in_req.password, &hashed_password_with_salt.0, &hashed_password_with_salt.1);

        if !verify_password_result.unwrap() {
            return Err(format!("Invalid password"));
        }

        let user = self.user_repo.log_in(log_in_req.email.clone(), hashed_password_with_salt.0).await?;

        let token = generate_token(self.jwt_secret_key.clone(), user).await?;

        Ok(token)
    }

}

File: internal/domain/usecase/user_usecase/log_in_usecase.go
package user_usecase

import (
	"blueberry_homework/dto/request"
	"blueberry_homework/dto/response"
	"blueberry_homework/utils/jwt"
	"errors"

	"golang.org/x/crypto/bcrypt"
)

// Login은 사용자 로그인을 처리합니다.
// false - 로그인 실패, true - 로그인 성공
func (u *UserUsecase) Login(req request.LoginRequest) (response.TokenResponse, error) {
	userId, err := u.repo.FindByEmail(req.Email)
	if err != nil {
		return response.TokenResponse{}, err
	}

	hashedPassword, err := u.repo.GetHashedPassword(req.Email)
	if err != nil {
		return response.TokenResponse{}, err
	}

	// 비번 검증
	err = bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(req.Password))
	if err != nil {
		return response.TokenResponse{}, errors.New("invalid password")
	}

	email, name, err := u.repo.GetTokenInfo(userId)
	if err != nil {
		return response.TokenResponse{}, err
	}

	token, err := jwt.GenerateToken(userId.String(), email, name)
	if err != nil {
		return response.TokenResponse{}, err
	}

	return response.TokenResponse{Token: token}, nil
}
